UseCases layout with associated components

wackerow opened the pull request:
## Description
Implements the UseCases layout, along with components used by each of these pages

- Implements `layouts/UseCases.tsx`, analogous to `templates/use-cases.tsx` from Gatsby repo
- Updated all image asset references for pages utilizing the `use-cases` layout
- Implemented `TableOfContents` component, which involved the addition of a markdown parser to generate the table of contents items.
- `generateTableOfContents` can be found in `src/components/TableOfContents/utils`, which takes a string of markdown content, parses out comments, and then filters down to only the header lines, and parses them into nested items to match the interface needed for the ToC component.

## Temporary changes that leave TODOs
- All i18n components have been disabled temporarily until i18n implementation ready
- `PageMetadata` component withheld until i18n implementation ready
- TableOfContents pointing to ids that are not yet being used in the actual header element (thus the links don't actually work)

## Preview URLs
- https://deploy-preview-33--ethereum-org-fork.netlify.app/defi
- https://deploy-preview-33--ethereum-org-fork.netlify.app/nft
- https://deploy-preview-33--ethereum-org-fork.netlify.app/dao
- https://deploy-preview-33--ethereum-org-fork.netlify.app/desci
- https://deploy-preview-33--ethereum-org-fork.netlify.app/refi
- https://deploy-preview-33--ethereum-org-fork.netlify.app/social-networks
- https://deploy-preview-33--ethereum-org-fork.netlify.app/decentralized-identity

wackerow committed:
initialize UseCasesLayout

wackerow committed:
Merge branch 'layout-templating' into use-cases

wackerow committed:
implement minimum viable UseCases layout

wackerow committed:
implement hero images for UseCases

wackerow committed:
implement ButtonDropdown for UseCases

required addition of matomo.ts util

wackerow committed:
implement TableOfContents

added util to parse heading IDs for their depth, titles and URLs. Will still need to adjust to make sure these URLs match the correct heading IDs in the DOM

wackerow committed:
enable refi page, disabled embedded image temporarily

temporarily removes error being caused by showing images

wackerow committed:
move TranslationKey to types.ts

fixes RandomAppList and UpgradeStatus—removed from tsconfig ignore

wackerow committed:
fix UpgradeTableOfContents

wackerow committed:
Merge branch 'dev' into use-cases

wackerow committed:
Merge branch 'dev' into use-cases

wackerow committed:
update GlossaryTooltip dependencies

wackerow committed:
enable GlossaryTooltip import in UseCases

sort import and components

wackerow committed:
update QuizWidget and dependencies

export with useCasesComponents

wackerow committed:
fix ToC header ID generation

moves call of generateTableOfContents to md.ts; add removeMarkdownComments to strip out commented out headers first

wackerow committed:
update interface for PageContent

wackerow committed:
fix refi image

wackerow committed:
clean up and add todos

wackerow committed:
Merge branch 'dev' into use-cases

pettinarip committed:
fix ignored button stories path

pettinarip committed:
fix button import

pettinarip commented on public/content/refi/refi-regenerative-economy-diagram.png:
Hmm what happened with this change?  I opened the file and it is empty 🤔

pettinarip commented on src/components/Glossary/GlossaryDefinition/index.tsx:
```diff
@@ -17,13 +17,16 @@ const GlossaryDefinition: React.FC<IProps> = ({ term, size = "md" }) => {
-      <OldHeading as="h3" lineHeight={1.4} id={term} {...headingStyles}>
-        <Translation id={`${term}-term`} />
-      </OldHeading>
+      <Heading as="h3" lineHeight={1.4} id={term} {...headingStyles}>
```
I appreciate these changes but I don't think we should include them in the scope of this PR and also in the migration since:
1. will make it more complex to keep in sync with the Gatsby repo since now we are adding new changes in the new repo that don't exist in the other repo
2. we might introduce styling bugs

Having said that, using the `Heading` component here will remove some default spacing that the `OldHeading` has.

wackerow replied:
Yeah was noticing some of the layout bugs as a result, will revert

pettinarip commented on src/components/Quiz/constants.ts:
```diff
@@ -0,0 +1,8 @@
+export const INITIAL_QUIZ = "what-is-ethereum"
+export const TOTAL_QUIZ_QUESTIONS_ANSWERED = 100000
+export const TOTAL_QUIZ_AVERAGE_SCORE = 67.4
+export const TOTAL_QUIZ_RETRY_RATE = 15.6
```
This was living in the `quizzes` page before, right?

wackerow replied:
@pettinarip It was actually in `/src/constants.ts`... Probably should just move this back to constants now that I'm thinking about it. This isn't a pattern we're using yet of nesting `constant.ts` files other than in the Simulator (which we could also consider relocating to the root file separately)

pettinarip commented on src/lib/types.ts:
```diff
@@ -70,3 +70,33 @@ export type Lang =
+export type TranslationKey = string
+
+/**
+ * Quiz data types
```
nice to have:

IMO types in this files are more generic and global.

The quizzes types are very specific, I'd place them in a new `types` file inside the `Quiz` folder.

wackerow replied:
Can't disagree... do we want to start bringing those types of adjustments into scope though? This is parity to Gatsby repo; I think I agree it's a nice to have if it happens, but wouldn't focus on this.

pettinarip commented on src/components/Quiz/utils.ts:
```diff
@@ -1,22 +1,21 @@
 
 export const getTotalQuizzesPoints = () =>
-  Object.values(allQuizzesData)
+  Object.values(allQuizzesData as RawQuizzes)
```
Not needed. `allQuizzesData` is already of type `RawQuizzes`

pettinarip commented on src/components/TableOfContents/TableOfContentsLink.tsx:
```diff
@@ -1,5 +1,5 @@
 import React from "react"
-import { Link as GatsbyLink } from "gatsby"
+// import { Link as GatsbyLink } from "gatsby"
```
lets remove this line as we are importing our custom `Link` component below.
```suggestion
```

pettinarip commented on src/components/TableOfContents/TableOfContentsMobile.tsx:
```diff
@@ -21,9 +20,6 @@ export interface IPropsTableOfContentsMobile {
 
 const Mobile: React.FC<IPropsTableOfContentsMobile> = ({ items, maxDepth }) => {
-  // TODO: Replace with direct token implementation after UI migration is completed
-  const lgBp = useToken("breakpoints", "lg")
```
Related to the previous comment about doing these changes here. I wouldn't do them here.

pettinarip commented on src/components/TableOfContents/index.tsx:
```diff
@@ -41,9 +37,6 @@ const TableOfContents: React.FC<IProps> = ({
   ...rest
 }) => {
-  // TODO: Replace with direct token implementation after UI migration is completed
-  const lgBp = useToken("breakpoints", "lg")
```
out of scope

pettinarip commented on tsconfig.json:
🔥

pettinarip commented on src/components/TableOfContents/utils.ts:
```diff
@@ -56,3 +57,68 @@ export const outerListProps: ListProps = {
+  )
+  depths.forEach((depth, i): void => {
+    if (depth > h) return
+    const headingItem: Item = parseHeadingToItem(headings[i])
```
not needed
```suggestion
    const headingItem = parseHeadingToItem(headings[i])
```

pettinarip commented on src/components/TableOfContents/utils.ts:
```diff
@@ -56,3 +57,68 @@ export const outerListProps: ListProps = {
+ * @param content Markdown content as a string (all lines)
+ * @returns List of `Item` objects parsed from the content, nested according to heading depth
+ */
+export const generateTableOfContents = (content: string): Array<Item> => {
```
Nice!

pettinarip commented on src/components/UpgradeTableOfContents.tsx:
```diff
@@ -87,19 +73,12 @@ const ItemsList: React.FC<IPropsItemsList> = ({ items, depth, maxDepth }) => {
-
+}
+const UpgradeTableOfContents: React.FC<IPropsToC> = ({ items, maxDepth = 1}) => {
+  console.log({yeah: items})
```
✂️

wackerow replied:
😆 oops

pettinarip commented on src/layouts/UseCases.tsx:
```diff
@@ -1,3 +1,448 @@
+}
+
+const Heading1 = (props: HeadingProps) => (
+  <Heading as="h1" {...commonHeadingProps} fontSize="2.5rem" {...props} />
```
We should keep using the `OldHeading` or set a `mb` to keep the spacing consistent.
![image](https://github.com/ethereum/ethereum-org-fork/assets/468158/10e4d568-552b-4504-8877-f914e201c763)

wackerow replied:
To prevent scope creep, will revert to use OldHeading (hah sorry, it's marked as deprecated so it gets crossed out in VSCode... hard to resist)

pettinarip commented on src/layouts/UseCases.tsx:
```diff
@@ -1,3 +1,448 @@
+  )
+}
+
+export const UseCasesLayout = ({ children, frontmatter, slug, tocItems }) => {
```
Lets define the types for the layout

wackerow replied:
We have `frontmatter, slug, tocItems` defined locally inside `md.ts` right now as `type Items`... I'd suggest we rename to `interface PageContent` and pull that into `@/lib/interfaces` so we can reuse in places like right here... will go ahead with that, but let me know if you disagree

pettinarip commented on src/layouts/UseCases.tsx:
```diff
@@ -1,3 +1,448 @@
+}
+
+export const UseCasesLayout = ({ children, frontmatter, slug, tocItems }) => {
+  if (!frontmatter)
+    throw new Error("UseCasesLayout missing expected `frontmatter` prop")
+  if (!frontmatter?.title)
+    throw new Error("Required `title` property missing for UseCasesLayout")
```
With the correct typing we shouldn't need these checks anymore

pettinarip commented on src/layouts/UseCases.tsx:
```diff
@@ -1,3 +1,448 @@
+      </Page>
+    </Box>
+  )
+}
```
Are there any significant changes you remember doing in this file? 

Asking just to keep track of those changes and keep them in mind for future repo syncs.

wackerow replied:
Main thing would probably be adjustments to the HeroContainer and the Image contained within when converting from GatsbyImage

pettinarip reviewed (changes requested):
Great job overall @wackerow 

Good to see that the ToC is now working as well 💪🏼

nhsz commented on src/components/TableOfContents/utils.ts:
@wackerow utils should be defined inside `src/lib/utils`

wackerow replied:
Okay... it was nested with the component in the Gatsby repo, are we trying to generally avoid doing that with this transition?

nhsz commented:
@wackerow noticing some visual differences (font weight, margins) vs our current design that apply to all added pages, eg: these are some buttons with missing margins on defi page (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)

<img width="1003" alt="Screen Shot 2023-10-04 at 12 11 48" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/198f75fb-1847-4107-924d-69f1d3e19794">

<img width="1021" alt="Screen Shot 2023-10-04 at 12 11 58" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/b7dacbca-09bd-4b2d-a347-f1febc99d5f0">

<img width="1008" alt="Screen Shot 2023-10-04 at 12 12 16" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/c8ff3780-7502-4a3e-9190-5b0476663452">

<img width="995" alt="Screen Shot 2023-10-04 at 12 12 27" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/f8c8f5f2-7364-4ba3-b271-82113c9245b1">

---

These are some differences in font weight, margins

### PREVIEW (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)

<img width="397" alt="1" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/8f90dc06-8978-43a5-9979-184a4ebdcc6b">

### PROD (https://ethereum.org/en/defi/)

<img width="381" alt="2" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/504aa552-2709-4879-9ddd-f41dcac873d9">

---

### PREVIEW (https://deploy-preview-33--ethereum-org-fork.netlify.app/defi)

<img width="961" alt="1b" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/6de7b27f-11a8-4e0d-8659-3422441ce871">

### PROD (https://ethereum.org/en/defi/)

<img width="990" alt="2b" src="https://github.com/ethereum/ethereum-org-fork/assets/948922/3dca7895-add0-4594-b3ec-cbecf3ebb9e1">

---

Maybe some could be related to recent changes pulled to theme @pettinarip?

wackerow committed:
move ToC utils and interfaces to src/lib

wackerow committed:
implement rehypeHeaderIds

adds custom header IDs to markdown heading elements

wackerow committed:
fix type issues for rehypeHeaderIds

wackerow committed:
add comments to and rename rehypeHeadingIds

wackerow commented on public/content/refi/refi-regenerative-economy-diagram.png:
File was not loading from `/public/use-cases/refi...png` so relocated next to markdown file (`/public/content/refi/refi...png`) which allowd it to load just fine

wackerow commented on src/components/Quiz/utils.ts:
@nhsz Similar to your comment about `TableOfContents/utils.ts` belonging in `/src/lib/utils/`, do we also want to relocate this set of utilities in the same manner? I moved the ToC utils to `/src/lib/utils/toc.ts` and could similarly move these to `/src/lib/utils/quiz.ts`. cc: @pettinarip

wackerow commented on src/components/UpgradeTableOfContents.tsx:
```diff
@@ -1,44 +1,13 @@
+import type { ToCItem } from "@/lib/interfaces"
+import { trimmedTitle } from "@/lib/utils/toc"
 
-const customIdRegEx = /^.+(\s*\{#([A-Za-z0-9\-_]+?)\}\s*)$/
-
-const slugify = (s: string): string =>
-  encodeURIComponent(String(s).trim().toLowerCase().replace(/\s+/g, "-"))
-
-const getCustomId = (title: string): string => {
-  const match = customIdRegEx.exec(title)
-  if (match) {
-    return match[2].toLowerCase()
-  }
-  console.warn("Missing custom ID on header: ", title)
-  return slugify(title)
-}
-
-const trimmedTitle = (title: string): string => {
-  const match = customIdRegEx.exec(title)
-  return match ? title.replace(match[1], "").trim() : title
```
Repeated functions... these are now shared from `/src/lib/utils/toc.ts`

wackerow committed:
Merge branch 'dev' into use-cases

wackerow committed:
implement suggestions from code review

wackerow committed:
import/export GlossaryDefinition for Static layout

used by /glossary

pettinarip committed:
ignore paths from netlify build

wackerow committed:
Merge pull request #39 from ethereum/netlify-ignore-files

Netlify ignore files

wackerow committed:
move Quiz constants back to @/lib/constants

wackerow committed:
Implement suggestion from code review

Co-Authored-By: Pablo Pettinari <pettinarip@gmail.com>

pettinarip commented on src/components/Quiz/utils.ts:
```diff
@@ -1,19 +1,17 @@
-
-import { CompletedQuizzes, QuizShareStats } from "../../types"
+import { getLocaleForNumberFormat } from "@/lib/utils/translations"
+import { CompletedQuizzes, type Lang, QuizShareStats } from "@/lib/types"
```
```suggestion
import type { CompletedQuizzes, Lang, QuizShareStats } from "@/lib/types"
```

pettinarip commented on src/components/TableOfContents/index.tsx:
```diff
@@ -13,18 +11,18 @@ import {
+import type { ToCItem } from "@/lib/interfaces"
 
-export { Item }
+export { ToCItem }
```
Wondering if this is even necessary now that we've moved it to `@/lib/interfaces`. I think we should consume it from there in all the other places.

wackerow replied:
Agree, this isn't a pattern we use anywhere else... Just did a search and doesn't look like we are actually this export in any other file (they all import from `@/lib/interfaces`), so I'll just remove this line.

pettinarip commented on src/lib/rehype/rehypeHeadingIds.ts:
```diff
@@ -0,0 +1,37 @@
+        value: trimmedTitle(lastChildValue),
+      }
+    }
+    // TODO: Add pseudo-element to heading to allow for anchor links
```
Is there a card to track this? I'm not sure. Happy to create one if it is missing.

wackerow replied:
Not yet... I see you added this to your summary list though, so yeah we'll make cards for those.

pettinarip commented on src/lib/rehype/rehypeHeadingIds.ts:
```diff
@@ -0,0 +1,37 @@
+ * The `getCustomId` Table of Contents utility functions is used to generate the `id`
+ * for each, allowing ToC links to match.
+ * The `trimmedTitle` function is used to remove any trailing `{#id}` from the heading
+ * @returns Plugin<[{}], Root>, a rehype plugin tht can be used in [...dynamic-md-pages].tsx
```
👍🏼

pettinarip commented on src/pages/[...slug].tsx:
```diff
@@ -72,7 +74,7 @@ export const getStaticPaths: GetStaticPaths = () => {
         params: {
           // Splitting nested paths to generate proper slug
-          slug: file.slug.split("/").slice(1),
+          slug: file.slug!.split("/").slice(1),
```
not needed anymore
```suggestion
          slug: file.slug.split("/").slice(1),
```

pettinarip commented on src/pages/[...slug].tsx:
```diff
@@ -86,6 +88,7 @@ export const getStaticProps: GetStaticProps<Props, Params> = async (
   const params = context.params!
   const markdown = getContentBySlug(params.slug.join("/"))
   const frontmatter = markdown.frontmatter as Frontmatter
+  const tocItems = markdown.tocItems as Array<ToCItem>
```
not needed anymore with the new `PageContent` type
```suggestion
  const frontmatter = markdown.frontmatter
  const tocItems = markdown.tocItems
```

wackerow replied:
Agree... Working on a more robust front matter typing schema where we may need to add something back here, but can remove at the moment.

pettinarip commented on src/lib/interfaces.ts:
```diff
@@ -0,0 +1,64 @@
+export interface PageContent {
+  slug: string
+  content: string
+  frontmatter: Frontmatter
```
Thought that we were going to define the base required fields for `Frontmatter`, like `title` and `description`. Are we going to do that in a different PR?

pettinarip replied:
Oh just saw your changes in the other PR related to this. Ignore.

pettinarip commented on src/lib/interfaces.ts:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Layout interface
+ */
+export interface PageContent {
```
Should we call this `MdPageContent`? since we will have react pages and those will not share the same props.

wackerow replied:
I'm down for that... I can address this separately when handing the front matter schema.

pettinarip replied:
Sounds good!

pettinarip reviewed (commented):
Good job @wackerow I've added a few minor comments but overall it looks great.

These are the remaining tasks that I can think of that were raised from this PR (and we should tackle in separate PRs):
- Fix content images sizes in mobile
- Fix hero image size to match the prod version
- Add an anchor image in each heading (inside the new rehype plugin)

Am I missing something else?

wackerow committed:
Apply suggestions from code review

Co-authored-by: Pablo Pettinari <pettinarip@gmail.com>

wackerow committed:
remove unused export

wackerow committed:
typo fix

wackerow commented:
That list looks good @pettinarip , and agree we can tackle separately. Similarly, if we want to rename PageContent to MdPageContent that's fine with me, but I can just address that separately as well.

Should be ready for final set of eyes before merging. cc: @nhsz

nhsz commented on src/lib/utils/toc.ts:
```diff
@@ -0,0 +1,125 @@
+ * @param mdContent
+ * @returns string with {#id} removed
+ */
+export const removeAnchorLinks = (mdContent: string) =>
```
@wackerow saw this util (`removeAnchorLinks`) was removed from `md.ts`, are we using it somewhere now? otherwise should be removed

wackerow replied:
@nhsz We are not. This is primarily handled with `parseHeadingToItem` now. I just added a commit to remove this function... 

(Also updated `trimmedTitle` to be called `parseToCTitle` to be more explicit about what it's doing, and added comments for it.)

I think this is the last TODO for this PR.

nhsz reviewed (approved):
@wackerow great job! Left a comment regarding `removeAnchorLinks` util. Be sure to track related tasks yet to be done in Notion, mentioned in the comments

wackerow committed:
update naming, remove deprecated

Change "trimmedTitle" to "parseToCTitle" and updated comments to provide clearer context on usage
